### ISSUE TITLE
An unhandled fatal exception has occurred.

### ISSUE THREAD

**Issue by igneousdog** [EDDiscovery/EDDiscovery#3100]
An unhandled fatal exception has occurred.

```
==== BEGIN ====
SQLLiteExtensions.SQLProcessingThreadException: attempt to write a readonly database
attempt to write a readonly database ---> System.Data.SQLite.SQLiteException: attempt to write a readonly database
attempt to write a readonly database
   at System.Data.SQLite.SQLite3.Reset(SQLiteStatement stmt)
   at System.Data.SQLite.SQLite3.Step(SQLiteStatement stmt)
   at System.Data.SQLite.SQLiteDataReader.NextResult()
   at System.Data.SQLite.SQLiteDataReader..ctor(SQLiteCommand cmd, CommandBehavior behave)
   at System.Data.SQLite.SQLiteCommand.ExecuteReader(CommandBehavior behavior)
   at System.Data.SQLite.SQLiteCommand.ExecuteNonQuery(CommandBehavior behavior)
   at SQLLiteExtensions.SQLExtRegister.PutSetting(String key, String sqlname, Object value) in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLRegister.cs:line 141
   at SQLLiteExtensions.SQLExtRegister.PutSetting[T](String key, T value) in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLRegister.cs:line 115
   at SQLLiteExtensions.SQLProcessingThread`1.Job`1.Exec() in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLProcessingThread.cs:line 68
   --- End of inner exception stack trace ---
   at SQLLiteExtensions.SQLProcessingThread`1.Job`1.Wait() in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLProcessingThread.cs:line 91
   at SQLLiteExtensions.SQLProcessingThread`1.Execute[T](Func`1 func, Int32 skipframes, Int32 warnthreshold) in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLProcessingThread.cs:line 239
   at SQLLiteExtensions.SQLProcessingThread`1.ExecuteWithDatabase[T](Func`2 func, Int32 warnthreshold) in C:\Code\Examples\EDDiscovery\BaseUtilities\SQLLite\SQLLite\SQLProcessingThread.cs:line 0
   at EliteDangerousCore.DB.SystemsDatabase.SetEDSMGalMapLast(DateTime value) in C:\Code\Examples\EDDiscovery\EliteDangerousCore\EliteDangerous\SystemDB\SystemsDatabase.cs:line 245
   at EDDiscovery.EDDiscoveryController.BackgroundWorkerThread() in C:\Code\Examples\EDDiscovery\EDDiscovery\EDDiscoveryControllerMain.cs:line 336
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
===== END =====
```

**Comment by Eahlstan** [comment]
Any chance you were adding or deleting grids from the DB while playing?

**Comment by klightspeed** [comment]
This looks like a race condition, where multiple threads were trying to do a write operation.

```cs
        public T WithReadWrite<T>(Func<T> action)
        {
            if (ReadOnly)
            {
                SetReadWrite();

                try
                {
                    return action();
                }
                finally
                {
                    SetReadOnly();
                }
            }
            else
            {
                return action();
            }
        }
```

That probably needs to use something like a semaphore, such that the first writer makes the database read-write as it enters, and the last writer makes the database read-only as it leaves.

**Comment by Eahlstan** [comment]
> This looks like a race condition, where multiple threads were trying to do a write operation.
> 
> ```cs
>         public T WithReadWrite<T>(Func<T> action)
>         {
>             if (ReadOnly)
>             {
>                 SetReadWrite();
> 
>                 try
>                 {
>                     return action();
>                 }
>                 finally
>                 {
>                     SetReadOnly();
>                 }
>             }
>             else
>             {
>                 return action();
>             }
>         }
> ```
> 
> That probably needs to use something like a semaphore, such that the first writer makes the database read-write as it enters, and the last writer makes the database read-only as it leaves.

Is this what was fixed in #3069 ?

**Comment by klightspeed** [comment]
> > This looks like a race condition, where multiple threads were trying to do a write operation.
> > ```cs
> >         public T WithReadWrite<T>(Func<T> action)
> >         {
> >             if (ReadOnly)
> >             {
> >                 SetReadWrite();
> > 
> >                 try
> >                 {
> >                     return action();
> >                 }
> >                 finally
> >                 {
> >                     SetReadOnly();
> >                 }
> >             }
> >             else
> >             {
> >                 return action();
> >             }
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That probably needs to use something like a semaphore, such that the first writer makes the database read-write as it enters, and the last writer makes the database read-only as it leaves.
> 
> Is this what was fixed in #3069 ?

It's possible it was related.
